pkg/network/protocols/postgres/telemetry.go

```diff
@@ -97,3 +98,131 @@ func (t *Telemetry) Count(tx *ebpf.EbpfEvent, eventWrapper *EventWrapper) {
+// return if nothing to add, in this case also metric group summary ignores unmodified metrics.
+func (t *PgKernelTelemetry) update(kernMsgCounts *pgebpf.PostgresKernelMsgCount, isTLS bool) {
+	if kernMsgCounts == nil {
+		return
+	}
+	if isEmptyCounts(kernMsgCounts) {
+		log.Debugf("postgres kernel telemetry empty.")
+		return
+	}
+	delta := t.getDiff(kernMsgCounts, isTLS)
+	if isEmptyCounts(delta) {
+		log.Debugf("postgres kernel telemetry, no difference with previous counters.")
+		return
+	}
+	t.pgMessagesCountNotFound.Add(int64(delta.Messages_not_found), isTLS)
+	t.pgMessagesCountAboveMax.Add(int64(delta.Messages_above_max), isTLS)
+	t.pgMessagesCountErrors.Add(int64(delta.Messages_count_err), isTLS)
+	for bucketIndex := range t.pgMessagesCountBuckets {
+		t.pgMessagesCountBuckets[bucketIndex].Add(int64(delta.Messages_count_buckets[bucketIndex]), isTLS)
+	}
+	if isTLS {
+		t.pgKernelMsgCountTlsPrev = *kernMsgCounts
+	} else {
+		t.pgKernelMsgCountPlainPrev = *kernMsgCounts
+	}
+	s := t.metricGroup.Summary()
+	log.Debugf("postgres kernel telemetry, isTLS=%t, summary: %s", isTLS, s)
+}
+
+func (t *PgKernelTelemetry) getDiff(kernMsgCounts *pgebpf.PostgresKernelMsgCount, isTLS bool) *pgebpf.PostgresKernelMsgCount {
+	if isTLS {
+		delta := subtractCounts(kernMsgCounts, t.pgKernelMsgCountTlsPrev)
+		return delta
+	} else {
+		delta := subtractCounts(kernMsgCounts, t.pgKernelMsgCountPlainPrev)
+		return delta
+	}
+}
+
+// subtractCounts create a new PostgresKernelMsgCount by subtracting the counts of 'b' from 'a'.
+// calculate only increasing counters, otherwise subtraction gives unreasonably large values for unsigned integers.
+func subtractCounts(a *pgebpf.PostgresKernelMsgCount, b pgebpf.PostgresKernelMsgCount) *pgebpf.PostgresKernelMsgCount {
+	res := &pgebpf.PostgresKernelMsgCount{}
+	if a.Messages_not_found >= b.Messages_not_found {
+		res.Messages_not_found = a.Messages_not_found - b.Messages_not_found
+	}
+	if a.Messages_above_max >= b.Messages_above_max {
+		res.Messages_above_max = a.Messages_above_max - b.Messages_above_max
+	}
+	if a.Messages_count_err >= b.Messages_count_err {
+		res.Messages_count_err = a.Messages_count_err - b.Messages_count_err
+	}
+	res.Messages_count_buckets = subtractArrays(a.Messages_count_buckets, b.Messages_count_buckets)
+	return res
+}
+
+func isEmptyCounts(p *pgebpf.PostgresKernelMsgCount) bool {
+	if p.Messages_not_found > 0 {
+		return false
+	}
+	if p.Messages_above_max > 0 {
+		return false
+	}
+	if p.Messages_count_err > 0 {
+		return false
+	}
+	if !arrayIsEmpty(p.Messages_count_buckets) {
+		return false
+	}
+	return true
+}
+
+// subtractArrays returns new array[] = a[] - b[]
+func subtractArrays(a, b [pgebpf.KerMsgCountNumBuckets]uint64) [pgebpf.KerMsgCountNumBuckets]uint64 {
+	var result [pgebpf.KerMsgCountNumBuckets]uint64
+	for i := 0; i < pgebpf.KerMsgCountNumBuckets; i++ {
+		if a[i] >= b[i] {
+			result[i] = a[i] - b[i]
+		}
+	}
+	return result
+}
```

Comment:
the documentation does not match the implementation
if a[i] is < b[i], you report `0` instead of `a[] - b[]`

Reply:
I added comment and changed code